File: _includes/SampleCode/SampleCode/Samples/Graph/PhotoUpload.cpp
#include "pch.h"
#include "FBReturnObject.h"

using namespace Windows::Foundation::Collections;
using namespace Windows::Storage;
using namespace Windows::Storage::Streams;
using namespace Windows::Storage::Pickers;
using namespace Platform;
using namespace concurrency;
using namespace winsdkfb;
using namespace winsdkfb::Graph;

namespace SampleCode
{
    void UploadPhoto()
    {
        StorageFile^ selectedPhoto;
        // Read image file into selectedPhoto
        FileOpenPicker^ fop = ref new FileOpenPicker();
        fop->ViewMode = PickerViewMode::Thumbnail;
        fop->SuggestedStartLocation = PickerLocationId::PicturesLibrary;
        fop->FileTypeFilter->Append(L".jpg");
        fop->FileTypeFilter->Append(L".png");
        create_task(fop->PickSingleFileAsync()).then([&](StorageFile^ storageFile)
        {
            selectedPhoto = storageFile;
            return storageFile->OpenReadAsync();
        }).then([=, &selectedPhoto](IRandomAccessStreamWithContentType^ stream)
        {
            // Create media stream
            FBMediaStream^ fbStream = ref new FBMediaStream(selectedPhoto->Name, stream);

            // Get active session
            FBSession^ sess = FBSession::ActiveSession;
            if (sess->LoggedIn)
            {

                PropertySet^ parameters = ref new PropertySet();
                // Set media stream
                parameters->Insert("source", fbStream);

                // Create Graph API path
                String^ graphPath = sess->User->Id + L"/photos";

                // Create a json class factory with a class (FBReturnObject class) that
                // can receive and parse the json response returned
                FBJsonClassFactory^ fact = ref new FBJsonClassFactory([](String^ JsonText) ->
                    Object^
                {
                    auto returnObject = ref new FBReturnObject();
                    returnObject->Id = Windows::Data::Json::JsonObject::Parse(JsonText)->GetNamedString("id");
                    return returnObject;
                });

                FBSingleValue^ sval = ref new FBSingleValue(graphPath, parameters, fact);
                create_task(sval->PostAsync()).then([=](FBResult^ result)
                {
                    if (result->Succeeded)
                    {
                        FBReturnObject^ response = static_cast<FBReturnObject ^>(result->Object);
                    }
                    else
                    {
                        // Posting failed
                    }
                });
            }
        });
    }
}

File: _includes/SampleCode/SampleCodeCs/Samples/Graph/PhotoUpload.cs
using System;
using Windows.Foundation.Collections;
using Windows.Storage;
using Windows.Storage.Streams;
using Windows.Storage.Pickers;
using Newtonsoft.Json;
using winsdkfb;
using winsdkfb.Graph;

namespace SampleCode
{
    public static partial class SampleFunc
    {
        static async void PhotoUpload()
        {
            // Read image file into selectedPhoto
            var fop = new FileOpenPicker();
            fop.ViewMode = PickerViewMode.Thumbnail;
            fop.SuggestedStartLocation = PickerLocationId.PicturesLibrary;
            fop.FileTypeFilter.Add(".jpg");
            fop.FileTypeFilter.Add(".png");
            StorageFile selectedPhoto = await fop.PickSingleFileAsync();
            // Create media stream
            IRandomAccessStreamWithContentType stream = await selectedPhoto.OpenReadAsync();
            FBMediaStream fbStream = new FBMediaStream(selectedPhoto.Name, stream);

            // Get current session
            FBSession sess = FBSession.ActiveSession;
            if (sess.LoggedIn)
            {
                // Get current user
                FBUser user = sess.User;

                PropertySet parameters = new PropertySet();
                // Set media stream
                parameters.Add("source", fbStream);

                // Set Graph api path
                string path = "/" + user.Id + "/photos";

                var factory = new FBJsonClassFactory(s => {
                    return JsonConvert.DeserializeObject<FBReturnObject>(s);
                });

                var singleValue = new FBSingleValue(path, parameters, factory);
                var result = await singleValue.PostAsync();
                if (result.Succeeded)
                {
                    var response = result.Object as FBReturnObject;
                }
                else
                {
                    // Posting failed
                }
            }
        }
    }
}